### ISSUE TITLE
Prevent setting observable effects to [[ArrayLength]]

### ISSUE THREAD

**Issue by leobalter** [tc39/ecma262#874]
While reviewing [#870](https://github.com/tc39/ecma262/pull/870/commits/34ca26bae59aa0c53b91da1247b3555c45871b44#r110283499), I've got a question on whether we should define accessing [[ArrayLength]] right after an observable step.

For `TypedArray#fill` we should move the `[[ArrayLength]]` access to run before `ToNumber(value)`

```
1. Let O be the this value.
2. Perform ? ValidateTypedArray(O).
4. Let len be O.[[ArrayLength]].
3. Let value be ? ToNumber(value).
5. Let relativeStart be ? ToInteger(start).
```

The suggestion above allows a implementation to define what is done to [[ArrayLength]] if gets a detached buffer.

See that `TypedArray#length` returns `0` if it has a detached buffer:

```
If IsDetachedBuffer(buffer) is true, return 0.
Let length be O.[[ArrayLength]].
```

---

[ValidateAtomicAccess](https://tc39.github.io/ecma262/#sec-validateatomicaccess) is the other case in the specs we can verify `[[ArrayLength]]` being accessed after an observable step where we do not check for a detached buffer.

```
ValidateAtomicAccess( typedArray, requestIndex )

1. Assert: typedArray is an Object that has a [[ViewedArrayBuffer]] internal slot.
2. Let accessIndex be ? ToIndex(requestIndex).
3. Let length be typedArray.[[ArrayLength]].
...
```

This one is tricky. Because `typedArray` may already be detached before the observable `ToIndex` operation.

My suggestion is to access the `length` property for this case, while ValidateAtomicAccess is not a direct TypedArray operation and it can't protect it.

This way we have may minimize the chances for weird effects:

 ```
ValidateAtomicAccess( typedArray, requestIndex )

1. Assert: typedArray is an Object that has a [[ViewedArrayBuffer]] internal slot.
2. Let accessIndex be ? ToIndex(requestIndex).
3. Let length be ? Get(typedArray, "length").
...
```

**Comment by anba** [comment]
> For TypedArray#fill we should move the [[ArrayLength]] access to run before ToNumber(value)

Will be fixed in #866 (ae46560691db80741514388c8698011bb005a551).

> ValidateAtomicAccess is the other case in the specs we can verify [[ArrayLength]] being accessed after an observable step where we do not check for a detached buffer.

See <https://github.com/tc39/ecma262/pull/870#discussion_r110324934>.

**Comment by leobalter** [comment]
That's it, thanks!
